### PR TITLE
Fixed how to capture filename of attachment. (Issue #703)

### PR DESCRIPTION
**Pull request by antonini**
Fixed how to capture filename of attachment. (Issue #703)

### CODE REVIEW

**Comment by karato** [comment]
@antonini Thanks for the pull request, let me find a reviewer..

**Comment by karato** [comment]
@ggajos review it

File: netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java [hunk 1]
```diff
@@ -80,12 +96,23 @@
...
+        final Request file = new RqMultipart.Smart(
             new RqMultipart.Base(req)
+        ).single("file");
+        final Matcher matcher = FILE_NAME_PATTERN.matcher(
+            new RqHeaders.Smart(
+                new RqHeaders.Base(file)
+            ).single("Content-Disposition")
+        );
+        if (!matcher.find()) {
+            throw new RsFailure("Filename was not provided");
```
Comment (ggajos) [inline netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java]: @antonini if this is new code, think we should test cover it

File: netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java [hunk 2]
```diff
@@ -80,12 +96,23 @@
 
     @Override
     public Response act(final Request req) throws IOException {
-        final RqMultipart.Smart multi = new RqMultipart.Smart(
+        final Request file = new RqMultipart.Smart(
             new RqMultipart.Base(req)
+        ).single("file");
+        final Matcher matcher = FILE_NAME_PATTERN.matcher(
```
Comment (ggajos) [inline netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java]: @antonini if this constant is used once and only inside this method, maybe we can inline?

File: netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java [hunk 3]
```diff
@@ -80,12 +96,23 @@
...
+            new RqHeaders.Smart(
+                new RqHeaders.Base(file)
+            ).single("Content-Disposition")
+        );
+        if (!matcher.find()) {
+            throw new RsFailure("Filename was not provided");
+        }
+        final String name = URLDecoder.decode(
+            matcher.group(FILE_NAME_POS),
```
Comment (ggajos) [inline netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java]: @antonini if this constant is used once and only inside this method, maybe we can inline?

File: netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java (hunk 2 above)
Comment (antonini) [inline netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java]: @ggajos The point is that this is a RegEx pattern, so it's better to compile it just once and run using compile pattern.

Comment (ggajos) [inline netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java]: @antonini I see, make sense

File: netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java (hunk 3 above)
Comment (antonini) [inline netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java]: @ggajos Fixed.

File: netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java (hunk 1 above)
Comment (antonini) [inline netbout-web/src/main/java/com/netbout/rest/bout/TkAttach.java]: @ggajos Created unit test.

**Comment by antonini** [comment]
@ggajos Fixed. Please review it.

**Comment by ggajos** [comment]
@antonini thanks

**Comment by ggajos** [comment]
@rultor merge please

**Comment by dmzaytsev** [comment]
@antonini the reporter of the original ticket has successfully upload a text file and failed with a pdf file. I'm not quite understand how this PR fix this bug...

**Comment by antonini** [comment]
@dmzaytsev the reporter tried online on snapshot or locally? I saw that this was not merged and not deployed.

If he tried locally, did he created an S3 account and configured your local netbout to use it? That was the only way I've found to test this.

All files that are larger than 50kb are stored on amazon. Even a text file larger than that are stored on amazon.

**Comment by dmzaytsev** [comment]
@antonini ask the reporter 
it's your ticket :)

**Comment by antonini** [comment]
@dmzaytsev Ok, I just asked you because I didn't saw any interaction of the reporter here and even on the original issue.

**Comment by antonini** [comment]
@super132 Did you tried on snapshot or locally? I saw that this was not merged and not deployed.

If locally, did you created an S3 account and configured your local netbout to use it? That was the only way I've found to test this.

All files that are larger than 50kb are stored on amazon. Even a text file larger than that are stored on amazon.

**Comment by super132** [comment]
@antonini I tried it on production environment only. I did not setup any local instance of netbout.

**Comment by antonini** [comment]
@super132 As you can see, the PR was not merged and not deploy test environment (http://snapshot.netbout.com/) neither production. Can you test it locally?

**Comment by super132** [comment]
@antonini I got this exception when I upload to test environment with a pdf file with size 79kb

```
java.io.IOException: com.amazonaws.services.s3.model.AmazonS3Exception: The specified bucket does not exist (Service: Amazon S3; Status Code: 404; Error Code: NoSuchBucket; Request ID: 3DB23CC8631CFFDE), S3 Extended Request ID: sCL0m0VAjfwm9F0w5E1bOexaDM4qMqmPVd2U7ejndWlzsKL8hn1tOpQ1hp2kSicg/m5UQ7pSpVE=
	at com.jcabi.s3.AwsOcket.write_aroundBody10(AwsOcket.java:200)
	at com.jcabi.s3.AwsOcket$AjcClosure11.run(AwsOcket.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
	at com.jcabi.aspects.aj.MethodLogger.wrap(MethodLogger.java:207)
	at com.jcabi.aspects.aj.MethodLogger.ajc$inlineAccessMethod$com_jcabi_aspects_aj_MethodLogger$com_jcabi_aspects_aj_MethodLogger$wrap(MethodLogger.java:1)
	at com.jcabi.aspects.aj.MethodLogger.wrapClass(MethodLogger.java:136)
	at com.jcabi.s3.AwsOcket.write(AwsOcket.java:184)
	at com.netbout.dynamo.DyAttachment.save(DyAttachment.java:280)
	at com.netbout.dynamo.DyAttachment.write_aroundBody10(DyAttachment.java:226)
	at com.netbout.dynamo.DyAttachment$AjcClosure11.run(DyAttachment.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
	at com.jcabi.aspects.aj.MethodLogger.wrap(MethodLogger.java:207)
	at com.jcabi.aspects.aj.MethodLogger.ajc$inlineAccessMethod$com_jcabi_aspects_aj_MethodLogger$com_jcabi_aspects_aj_MethodLogger$wrap(MethodLogger.java:1)
	at com.jcabi.aspects.aj.MethodLogger.wrapClass(MethodLogger.java:136)
	at com.netbout.dynamo.DyAttachment.write(DyAttachment.java:191)
	at com.netbout.cached.CdAttachment.write_aroundBody18(CdAttachment.java:99)
	at com.netbout.cached.CdAttachment$AjcClosure19.run(CdAttachment.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
	at com.jcabi.aspects.aj.MethodLogger.wrap(MethodLogger.java:207)
	at com.jcabi.aspects.aj.MethodLogger.ajc$inlineAccessMethod$com_jcabi_aspects_aj_MethodLogger$com_jcabi_aspects_aj_MethodLogger$wrap(MethodLogger.java:1)
	at com.jcabi.aspects.aj.MethodLogger.wrapClass(MethodLogger.java:136)
	at com.netbout.cached.CdAttachment.write(CdAttachment.java:99)
	at com.netbout.email.EmAttachment.write_aroundBody10(EmAttachment.java:91)
	at com.netbout.email.EmAttachment$AjcClosure11.run(EmAttachment.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
	at com.jcabi.aspects.aj.MethodLogger.wrap(MethodLogger.java:207)
	at com.jcabi.aspects.aj.MethodLogger.ajc$inlineAccessMethod$com_jcabi_aspects_aj_MethodLogger$com_jcabi_aspects_aj_MethodLogger$wrap(MethodLogger.java:1)
	at com.jcabi.aspects.aj.MethodLogger.wrapClass(MethodLogger.java:136)
	at com.netbout.email.EmAttachment.write(EmAttachment.java:91)
	at com.netbout.rest.bout.TkAttach.act(TkAttach.java:105)
	at com.netbout.rest.bout.FkBout$3.act(FkBout.java:111)
	at com.netbout.rest.bout.FkBout$2.act(FkBout.java:89)
	at org.takes.facets.fork.FkRegex.route(FkRegex.java:189)
	at com.netbout.rest.bout.FkBout.route(FkBout.java:81)
	at com.netbout.rest.bout.FkBout.access$000(FkBout.java:52)
	at com.netbout.rest.bout.FkBout$1.route(FkBout.java:65)
	at org.takes.facets.fork.FkWrap.route(FkWrap.java:61)
	at org.takes.facets.fork.TkFork.act(TkFork.java:99)
	at org.takes.tk.TkWrap.act(TkWrap.java:59)
	at org.takes.facets.fork.FkRegex$1.act(FkRegex.java:152)
	at org.takes.facets.fork.FkRegex.route(FkRegex.java:189)
	at org.takes.facets.fork.TkFork.act(TkFork.java:99)
	at com.netbout.rest.FkRegistered$1.route(FkRegistered.java:70)
	at org.takes.facets.fork.FkWrap.route(FkWrap.java:61)
	at org.takes.facets.fork.TkFork.act(TkFork.java:99)
	at org.takes.facets.fork.FkAuthenticated.route(FkAuthenticated.java:82)
	at org.takes.facets.fork.TkFork.act(TkFork.java:99)
	at org.takes.facets.auth.TkAuth.act(TkAuth.java:107)
	at org.takes.facets.auth.TkAuth.act(TkAuth.java:87)
	at org.takes.tk.TkWrap.act(TkWrap.java:59)
	at org.takes.facets.forward.TkForward.act(TkForward.java:66)
	at org.takes.facets.fallback.TkFallback.route(TkFallback.java:78)
	at org.takes.facets.fallback.TkFallback.access$000(TkFallback.java:48)
	at org.takes.facets.fallback.TkFallback$1.act(TkFallback.java:60)
	at org.takes.tk.TkWrap.act(TkWrap.java:59)
	at org.takes.tk.TkWrap.act(TkWrap.java:59)
	at org.takes.facets.flash.TkFlash.act(TkFlash.java:98)
	at org.takes.tk.TkMeasured$1.act(TkMeasured.java:65)
	at org.takes.tk.TkWrap.act(TkWrap.java:59)
	at org.takes.tk.TkVersioned$1.act(TkVersioned.java:70)
	at org.takes.tk.TkWrap.act(TkWrap.java:59)
	at org.takes.tk.TkWithHeaders$1.act(TkWithHeaders.java:69)
	at org.takes.tk.TkWrap.act(TkWrap.java:59)
	at org.takes.tk.TkWrap.act(TkWrap.java:59)
	at org.takes.http.BkBasic.print(BkBasic.java:99)
	at org.takes.http.BkBasic.accept(BkBasic.java:77)
	at org.takes.http.BkSafe.accept(BkSafe.java:58)
	at org.takes.http.BkTimeable.accept(BkTimeable.java:94)
	at org.takes.http.BkParallel$1.run(BkParallel.java:97)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.amazonaws.services.s3.model.AmazonS3Exception: The specified bucket does not exist (Service: Amazon S3; Status Code: 404; Error Code: NoSuchBucket; Request ID: 3DB23CC8631CFFDE), S3 Extended Request ID: sCL0m0VAjfwm9F0w5E1bOexaDM4qMqmPVd2U7ejndWlzsKL8hn1tOpQ1hp2kSicg/m5UQ7pSpVE=
	at com.amazonaws.http.AmazonHttpClient.handleErrorResponse(AmazonHttpClient.java:1077)
	at com.amazonaws.http.AmazonHttpClient.executeOneRequest(AmazonHttpClient.java:725)
	at com.amazonaws.http.AmazonHttpClient.executeHelper(AmazonHttpClient.java:460)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:295)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:3714)
	at com.amazonaws.services.s3.AmazonS3Client.putObject(AmazonS3Client.java:1445)
	at com.jcabi.s3.AwsOcket.write_aroundBody10(AwsOcket.java:188)
	... 71 more
```

When I try uploading a text file (.cpp) file with 2KB, I can upload it successfully without exception.

Hope it helps

**Comment by antonini** [comment]
@super132 As you can see, the bucket you informed does not exists.

Did you configured those three keys on `netbout-web/src/test/resources/META-INF/MANIFEST.MF` ?

    Netbout-S3Key: test
    Netbout-S3Secret: test
    Netbout-S3Bucket: test

**Comment by dmzaytsev** [comment]
@antonini take a look  #717

**Comment by antonini** [comment]
@dmzaytsev I didn't touch on that API. My change will not fix the problem that is happening when using REST.

**Comment by dmzaytsev** [comment]
@super132 I just successfully uploaded a pdf file on [production environment](http://www.netbout.com)
please try to upload this file http://www.publishers.org.uk/_resources/assets/attachment/full/0/2091.pdf on production

**Comment by dmzaytsev** [comment]
@yegor256 could you please upload the [file](https://static.googleusercontent.com/media/source.android.com/ru//devices/tech/security/reports/Google_Android_Security_2014_Report_Final.pdf) on [production](http://www.netbout.com) and share the exception from the server log here ?

**Comment by yegor256** [comment]
@dmzaytsev there are no logs :( I just see "bad request" white page and that's it

**Comment by antonini** [comment]
@dmzaytsev I receive this message when I try to upload this on production.

error

**Comment by dmzaytsev** [comment]
@antonini you forgot to set the attachment name..
we can upload http://www.publishers.org.uk/_resources/assets/attachment/full/0/2091.pdf
but another file https://static.googleusercontent.com/media/source.android.com/ru//devices/tech/security/reports/Google_Android_Security_2014_Report_Final.pdf causes "bad request"

**Comment by dmzaytsev** [comment]
@antonini If you already have properly configured the local instance
could you try to upload these files on local?

**Comment by antonini** [comment]
@dmzaytsev I tryed to upload this file `http://www.publishers.org.uk/_resources/assets/attachment/full/0/2091.pdf` using `example.pdf` as a name.

I´ll try to upload those files on my local environment.

**Comment by antonini** [comment]
@dmzaytsev Sorry.. clicked on "close" by mistake! :(

**Comment by antonini** [comment]
@dmzaytsev I think this pull request is not going to solve whole problem.This is solving just a problem that I found about getting file name. I'm investigating the problem and I'm pretty sure that we have a problem inside takes project. I'll confirm that.

**Comment by dmzaytsev** [comment]
@antonini  what a Takes bug you mean?
btw how we can get an invalid header `Content-Disposition` in real life? A browser makes the header, isn't it?

**Comment by antonini** [comment]
@dmzaytsev Is not a invalid header that the browser generated. Is a situation that is not supported by Takes. I'm reading this [document](http://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html), the section about "Chunked Transfer Coding" and when I debug RqLive I got a "trailer" message and a end of message that is not supported by takes.

**Comment by dmzaytsev** [comment]
@antonini feel free submit a bug in Takes, then create a puzzle to upgrade Netbout to new Takes version.
And we'll merge this PR

**Comment by antonini** [comment]
@dmzaytsev Issue [takes/#372](https://github.com/yegor256/takes/issues/372) was created due this situation. Where do I put a puzzle? On pom file?

**Comment by dmzaytsev** [comment]
@antonini sure, you can put it into pom file

**Comment by antonini** [comment]
@dmzaytsev @ggajos Puzzle added.

**Comment by dmzaytsev** [comment]
@antonini thanks

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by karato** [comment]
@ypshenychka review this ticket please, for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by karato** [comment]
@rultor deploy

**Comment by ypshenychka** [comment]
@karato Quality is good here.

**Comment by karato** [comment]
> @karato Quality is good here.

@ypshenychka thanks!

**Comment by karato** [comment]
@ggajos I added 10 mins to @ypshenychka (for QA review) in transaction 62532989; since quality is good, I just added 10 mins to @dmzaytsev (our architect) in transaction 62532999; Many thanks! **17 mins** were added to your account in Transaction ID `AP-87A07295BY113221N` (task took 336 hours and 33 mins); review comments (c=2) added as a bonus; +17 to your rating, your total score is [+566](http://www.netbout.com/b/35094?open=rating)

**Pull request merged**